### PR TITLE
ENH: PMPS utilities.

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

Add some functions for predicting PMPS behavior and producing the eV bitmasks:
- `get_bitmask`: given an upper/lower of a range, return the bitmask that matches that range.
- `check_bitmask`: given a bitmask and an eV value, return True if the energy is allowed.
- `check_actual_range`: given an upper/lower requested range, return the actual effective range based on the precision of the bitmask values. (we can't get exactly the range we want due to the bitmask architecture)

Note that the eV ranges used are going to be updated in pmps lib soon and will need to be updated again.

Also establishes a new doctr deploy key (the old needed to be revoked).

## Motivation and Context

This code will a useful utility for understanding the PMPS and for predicting its behavior.
The `get_bitmask` function will be ported to TwinCAT in the PMPS library as a PLC-side utility, and on that end will help keep the PLC code up-to-date as the PMPS bitmask definitions change.
https://jira.slac.stanford.edu/projects/LCLSPC/issues/LCLSPC-121

## How Has This Been Tested?

Added appropriate unit tests.

## Where Has This Been Documented?

Added docstrings and an automodule page.

### CODE REVIEW

File: pcdscalc/pmps.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        Custom boundaries to use instead of the default soft-xray
+        or hard-xray lines. Useful for testing.
+
+    Returns
+    -------
+    bitmask: int
+    """
+    # Be lenient on input args
+    if upper < lower:
+        lower, upper = upper, lower
```
Comment (klauer) [inline pcdscalc/pmps.py:135]: But maybe not too lenient;
`assert lower >= 0 and upper >= 0`?

File: pcdscalc/pmps.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+            if lower <= prev and upper >= ev:
+                bitmask += 2**bit
+            prev = ev
+        return bitmask
+
+    # An exclusion range is just two inclusion ranges
+    else:
+        lower_range = get_bitmask(0, lower, True, line, bounds=bounds)
+        upper_range = get_bitmask(upper, math.inf, True, line, bounds=bounds)
+        return lower_range | upper_range
```
Comment (klauer) [inline pcdscalc/pmps.py:152]: Neat usage here! 

One thought:
* `get_bitmask(allow=True)` is inclusive at its bounds `[lower, upper]`.
* Noting that I haven't done the math here so I could easily be off-base -- would this let through `lower` and `upper`? And if so, is that intentional? The docstring implies they should be excluded.

**Review by klauer — APPROVED** [review]
Looks good to me! One of those things you probably wish you had months ago 😄 

It'd also be cool if you had a "describe_bitmask" function that said "Bit 1: 0 [ev_low, ev_high] disallowed; Bit 2: 1 ..." or something like that.

File: pcdscalc/pmps.py (hunk 2 above)
Comment (ZLLentz) [reply pcdscalc/pmps.py:152]: `get_bitmask(allow=True)` is intended to exclude the endpoints unless they match an eV boundary exactly.

I'm going to talk through the test cases to make sure they all make sense and I didn't miss an edge case (and because this is confusing):

```
     (0, 15.5, True,     0b0000_0000_0000_0000_0111_1111_1111_1111),
     (15.5, 100, True,   0b1111_1111_1111_1111_0000_0000_0000_0000),
     (14.5, 21.5, True,  0b0000_0000_0001_1111_1000_0000_0000_0000),
     (14.5, 21.5, False, 0b1111_1111_1100_0000_0011_1111_1111_1111),
     (15, 20, True,      0b0000_0000_0000_1111_1000_0000_0000_0000)
```
```
boundaries = [1, 2, 3, 4 ... 32]
```

1. Allow between 0 and 15, exclude the 15.5 point because we can't allow points like 15.6 and they share a range. Therefore, enable bits 1 through 15 (bit 1 allows 0 to 1, bit 15 allows 14 to 15)
2. Allow between 16 and 32, exclude the 15.5 point because we can't allow points like 15.4 and they share a range. Cut off at 32 for the top of the bitmask range. Therefore, enable bits 17 through 32 (bit 17 allows 16 to 17)
3. Allow between 15 and 21, exclude the exact points because we can't allow any points outside the range. Enable bits 16 to 21 (bit 16 allows 15 to 16, bit 21 allows 20 to 21)
4. Allow between 0 and 14 and between 22 and 32. Turn off bits 15 to 22 (bit 15 allows 14 to 15, bit 22 allows 21 to 22). Exclude the boundary bits for the same reasonings as above.
5. Allow exactly between 15 and 20. Turn on bits 16 to 20 (bit 16 allows 15 to 16, bit 20 allows 19 to 20)

Now, perhaps I'm missing the case of an exact range with False? Perhaps:
```
     (15, 20, False,     0b1111_1111_1111_0000_0111_1111_1111_1111)
```
Which turns off bits 16 to 20 (bit 16 allows 15 to 16, bit 20 allows 19 to 20)
Checking the algorithm:
```
In [8]: bin(get_bitmask(15, 20, False, 'tst', list(range(1,33))))
Out[8]: '0b11111111111100000111111111111111'
# 0b1111_1111_1111_0000_0111_1111_1111_1111
```
Ok, I think I'm satisfied with that one.

Is there some other case I'm missing?

File: pcdscalc/pmps.py (hunk 1 above)
Comment (ZLLentz) [reply pcdscalc/pmps.py:135]: As a warning to a wayward user who puts in a negative, because it's probably in error? I was being lenient and just accepting it and calculating appropriately, e.g. `lower, upper = -10, -1` would give an all zero bitmask.

I do see the value in raising an exception here for users who accidentally pass a negative in.

**Comment by ZLLentz** [comment]
I'll make a describe_bitmask function, it's easy enough to do.

File: pcdscalc/pmps.py (hunk 2 above)
Comment (klauer) [reply pcdscalc/pmps.py:152]: The only edge cases seem to be where they are on the exact boundary, which you've described.

Reading through your logic, it makes enough sense to me to be confident you've covered all the bases - thanks for that. 👍

**Comment by ZLLentz** [comment]
```
In [1]: from pcdscalc.pmps import get_bitmask, describe_bitmask

In [2]: bitmask = get_bitmask(1400, 3000, True, 'KFE')

In [3]: describe_bitmask(bitmask, 'KFE')
Bit  1: 0 (     0,    250) disallowed
Bit  2: 0 (   250,    270) disallowed
Bit  3: 0 (   270,    350) disallowed
Bit  4: 0 (   350,    400) disallowed
Bit  5: 0 (   400,    450) disallowed
Bit  6: 0 (   450,    540) disallowed
Bit  7: 0 (   540,    850) disallowed
Bit  8: 0 (   850, 1150.0) disallowed
Bit  9: 0 (1150.0, 1250.0) disallowed
Bit 10: 0 (1250.0, 1450.0) disallowed
Bit 11: 1 (1450.0, 1550.0) allowed
Bit 12: 1 (1550.0, 1650.0) allowed
Bit 13: 1 (1650.0, 1750.0) allowed
Bit 14: 1 (1750.0, 1820.0) allowed
Bit 15: 1 (1820.0, 2000.0) allowed
Bit 16: 1 (2000.0, 2300.0) allowed
```

File: pcdscalc/pmps.py (hunk 2 above)
Comment (ZLLentz) [reply pcdscalc/pmps.py:152]: :+1: I'm going to add most of that text as a block comment in the test suite

**Comment by klauer** [comment]
Awesome!

File: pcdscalc/pmps.py (hunk 1 above)
Comment (klauer) [reply pcdscalc/pmps.py:135]: Yeah, I think mainly for users passing in calculated ranges and messing up somehow. Doesn't look like it'd really affect the logic, of course.
